### ISSUE TITLE
[MNG-6370] ConcurrencyDependencyGraph#getNumberOfBuilds() does not remove finished projects from unfinished ones

### ISSUE THREAD

**Issue by jira-importer** [apache/maven#7478]
**[Sylwester Lachiewicz](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=slachiewicz)** opened **[MNG-6370](https://issues.apache.org/jira/browse/MNG-6370?redirect=false)** and commented

Fix _ConcurrencyDependencyGraph.getUnfinishedProjects_

```java
private final HashSet<MavenProject> finishedProjects = new HashSet<>();

/**
 * @return set of projects that have yet to be processed successfully by the build.
 */
public Set<MavenProject> getUnfinishedProjects()
{
 Set<MavenProject> unfinished = new HashSet<>( projectBuilds.getProjects() );
 unfinished.remove( finishedProjects );
 return unfinished;
}
```

replace _remove()_ with _removeAll()_

---

**Affects:** 3.5.2, 3.5.3

**Remote Links:**
- [GitHub Pull Request #161
](https://github.com/apache/maven/pull/161)

0 votes, 5 watchers

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

The current code will lead to a `ClassCastException`, right? If so, why wasn't this raised way earlier?

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

At least all ITs pass here.

**Comment by jira-importer** [comment]
**[Sylwester Lachiewicz](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=slachiewicz)** commented

No, will just return not modified array. As far i can check there was not calls to getUnfinishedProject() in Maven sources.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

You are right, checked the impl. No CCEs are thrown.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Fixed with [faa9ef0cd3de27d31064f88cd74140516b3569a3](https://git-wip-us.apache.org/repos/asf?p=maven.git;a=commit;h=faa9ef0cd3de27d31064f88cd74140516b3569a3).
